org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/text/correction/SourceAssistProcessor.java

```diff
@@ -552,14 +556,32 @@ private Optional<Either<Command, CodeAction>> getGenerateDelegateMethodsAction(C
 
 	private Optional<Either<Command, CodeAction>> addFinalModifierWherePossibleAction(IInvocationContext context) {
 		IProposableFix fix = (IProposableFix) VariableDeclarationFixCore.createCleanUp(context.getASTRoot(), true, true, true);
+		return getFinalModifierWherePossibleAction(context, fix, ActionMessages.GenerateFinalModifiersAction_label, JavaCodeActionKind.SOURCE_GENERATE_FINAL_MODIFIERS);
+	}
 
+	private Optional<Either<Command, CodeAction>> addFinalModifierWherePossibleQuickAssist(IInvocationContext context) {
+		// find fully covered nodes
+		String actionMessage = ActionMessages.GenerateFinalModifiersAction_selectionLabel;
+		IProposableFix fix = (IProposableFix) VariableDeclarationFixCore.createChangeModifierToFinalFix(context.getASTRoot(), QuickAssistProcessor.getFullyCoveredNodes(context, context.getCoveringNode()).toArray(new ASTNode[0]));

```

Comment:
I noticed if I select multiple field declarations, I don't see the quick assist to change all modifiers.

```
public class Item {
    private String label;
    private float price;
}
```

It works just fine for local variable declarations. `QuickAssistProcessor.getFullyCoveredNodes(..)` finds the multiple field declarations, but `VariableDeclarationFixCore.createChangeModifierToFinalFix(..)` is what fails to generate the fix.

Reply:
Right. That makes sense. I think what confused me was that it's possible to add the modifier individually :

![final-modifier-uninit-field](https://user-images.githubusercontent.com/1417342/174863081-870230af-892e-48f2-ae81-da74fc1b4b9d.gif)

If a local variable (not a field) has no initializer I also noticed it work, so would that be a bug as well ?

Reply:
Yeah, that should be a bug, I'll take a look.

Reply:
I found if we directly pass a `VariableDeclarationFragment` to `VariableDeclarationFixCore.createChangeModifierToFinalFix()`, it will just ignore the initialization check, but for a `FieldDeclaration`, it will check its initializer as expected. So we should do this ourselves. 

I have fix this via https://github.com/eclipse/eclipse.jdt.ls/pull/2134/commits/db41c99115a9ebbecee05cf85d3e62fb121867f7 and added some tests.

Reply:
Anything we can do about :

![final-modifier-uninit-bug2](https://user-images.githubusercontent.com/1417342/175127655-6fa77ae1-8d70-4d5f-9f78-131af0814699.gif)

Reply:
we should the similar check for the selected nodes (now inferred node only) since `VariableDeclarationFixCore.createChangeModifierToFinalFix()` will not check the initializer. Let me add it and add the related test case.

Reply:
As you mention, the problem is in `VariableDeclarationFixCore#createChangeModifierToFinalFix(..)` . When it calls `VariableDeclarationFinder`, it doesn't check that local variables have an initializer, in `visit(VariableDeclarationStatement node)` -> `handleFragments` -> `canAddFinal` .

As a result, I still see things like :

![final-modifier-uninit-bug3](https://user-images.githubusercontent.com/1417342/175394882-a7aefc63-0c22-43f5-aa16-3e3f65540f70.gif)
 
I'm also still seeing the behaviour from https://github.com/eclipse/eclipse.jdt.ls/pull/2134/files#r904186818 with your recent changes.

Reply:
ah, I'm sorry that I didn't push a commit yesterday since I'm still finding a way to solve this :). 

Besides, I also find an interesting thing, the language server will not report an error if a local final variable has not been initialized, but will report for a field. e.g.,
```
public class AddModifier {
	private final String label1; // jdtls will report The blank final field label1 may not have been initialized
	public void test() {
		final String label2; // jdtls only reports The value of the local variable label2 is not used
	}
}
```

So it might explain why the initializer check happens in `fieldCanBeFinal()` 
 but not `canAddFinal()`, you can also find related code here. https://github.com/eclipse-jdt/eclipse.jdt.ui/blob/567c1546ee76fbbd4a50ee6b0de8e046a19356fe/org.eclipse.jdt.core.manipulation/core%20extension/org/eclipse/jdt/internal/corext/fix/VariableDeclarationFixCore.java#L187-L196

Coming back to the eclipse implementation, the change modifiers to final quick assist also appear for both label1(will cause an error) and label2(will not cause an error) in my example code. Anyway, showing the quick assist for label1 does not make sense to me since it will cause an error.

So our problems here are:

- if the compiler reports are correct, should we only show the quick assist for label2?
- if we can check the initializer ourselves, should we just do this check in jdtls, and create a new bug in JDT side, then remove this temporary check after the fix is available in upstream JDT?

Reply:
To answer your question above, the `final` modifier on a field, requires that it be initialized as part of the declaration statement. That's why there is an error. When the `final` keyword is added to a local variable that same restriction does not apply. One can still initialize it anywhere else in the scope as long as it happens only once. So I would say the `Add final modifier for ${var}` should always be available on a local variable. I think the logic you added to check there's an initializer isn't needed. Sorry about the confusion.

For a field, the behaviour seems fine to me.

I think with that fixed, this should be ready to merge.

Reply:
Get it. So I should remove the check of initializer here and open a ticket on JDT side?

Reply:
For now let's just mimic the behaviour of VariableDeclarationFixCore as it may be doing the right thing. I think there's quite a few things that determine whether to show modifiers or not (initializer, local vs. field, visibility, etc.).

So if you make a selection that fully covers a node (eg local variable, or field) and the quick assist for "Change modifiers to final" shows up for that selection, then the "Add final modifier for ${var}" should also show up when you simply place the cursor on the node and trigger quick-assist that way.

If we can just be consistent in that sense, then I think that's good enough for now.

Reply:
Sure. I have changed the behavior and added the related test cases. @rgrunber  Could you please check if there is any case I'm missing?